google-cloud-storage/src/main/java/com/google/cloud/storage/NotificationInfo.java

```diff
@@ -0,0 +1,342 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.storage;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.pathtemplate.PathTemplate;
+import com.google.api.services.storage.model.Notification;
+import com.google.common.base.Function;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Google Storage Notification metadata;
+ *
+ * @see <a href="https://cloud.google.com/storage/docs/concepts-techniques#concepts">Concepts and
+ *     Terminology</a>
+ */
+public class NotificationInfo implements Serializable {
+
+  private static final long serialVersionUID = 5725883368559753810L;
+  private static final PathTemplate PATH_TEMPLATE =
+      PathTemplate.createWithoutUrlEncoding("projects/{project}/topics/{topic}");
+
+  public enum PayloadFormat {
+    JSON_API_V1,
+    NONE
+  }
+
+  static final Function<Notification, NotificationInfo> FROM_PB_FUNCTION =
+      new Function<Notification, NotificationInfo>() {
+        @Override
+        public NotificationInfo apply(Notification pb) {
+          return NotificationInfo.fromPb(pb);
+        }
+      };
+  static final Function<NotificationInfo, Notification> TO_PB_FUNCTION =
+      new Function<NotificationInfo, Notification>() {
+        @Override
+        public Notification apply(NotificationInfo NotificationInfo) {
+          return NotificationInfo.toPb();
+        }
+      };
+  private final String generatedId;
+  private final String topic;
+  private final List<String> eventTypes;
+  private final Map<String, String> customAttributes;
+  private final PayloadFormat payloadFormat;
+  private final String objectNamePrefix;
+  private final String etag;
+  private final String selfLink;
+
+  /** Builder for {@code NotificatioInfo}. */
+  public abstract static class Builder {
+    Builder() {}
+
+    public abstract Builder setGeneratedId(String generatedId);
+
+    public abstract Builder setSelfLink(String selfLink);
+
+    public abstract Builder setTopic(String topic);
+
+    public abstract Builder setPayloadFormat(PayloadFormat payloadFormat);
+
+    public abstract Builder setObjectNamePrefix(String objectNamePrefix);
+
+    public abstract Builder setEventTypes(Iterable<String> eventTypes);
+
+    public abstract Builder setEtag(String etag);
+
+    public abstract Builder setCustomAttributes(Map<String, String> customAttributes);
+
+    /** Creates a {@code NotificationInfo} object. */
+    public abstract NotificationInfo build();
+  }
+
+  public static final class BuilderImpl extends Builder {
+
```

Comment:
Why do you need ` public abstract class Builder` and `public static final class BuilderImpl extends Builder`?
I guess just `public static final class Builder` will be enough

Reply:
Overuse of abstract classes and interfaces is why CLIRR is giving you troubles. Prefer concrete classes.